### DISCUSSION TITLE
Cue Point Question

### DISCUSSION THREAD

**Discussion by wppronto** [AzuraCast/AzuraCast#6978] [Q&A]
While I'm hoping to use the new autocue functionality to automate transitions, I'm not having good luck with it. So I may need to disable it and just use cue points on every song.

I don't want crossfades, I want radio-style segues. So I know that I need to add a face-in value of 0 to every file (unless there is a reason for it to fade in). And, at the end, I don't want to add any artificial fading in most cases, I just want the record to fade as it normally does, but I want to specify when the next song starts.

My question: Is it necessary to set a fade-out value, or will the overlap value simply indicate when to start the next song?

Thanks!!!

**Comment by unudelaradio** [comment]
Hello!

After the latest LS updates the secret stays in the AutoDJ Settings for Crossfade Duration (Seconds). That value you set there will be the general FADE duration (fade in & fade out) and general OVERLAP duration for all the tracks that **don't** have custom cue points.

- If you want all your tracks to start with no fade-in, yes, you should set a 0 value for each and every track.
- If you want to define where next track starts, yes, you should set custom overlaps for all your tracks, otherwise the next track will start on the last 2 seconds of the current song, if Crossfade Duration in set to 2 seconds in AutoDJ Settings. And so on for every other value you set in AutoDJ Settings. This AutoDJ default value will work as a fallback for all tracks that **don't** have custom cue points. 

But with endings is a little bit more complicated:

If you want all your tracks to end with no fade out... well... a 0 fade-out value for all tracks is not the way of doin' it. That will cut off all the songs at the start of your custom overlaps. A 0 or 0.1 fade out value will indicate that you want an extreme fade out effect, a cut, at the start of overlap.

As i noticed, Overlaps and Fade Outs always stick together in current Azuracast Version. At the start of every overlap (custom or not) a fade out will be applied. But you can choose: if this fade out is a custom one... or the default one in AutoDJ Settings.

As long you have 0 fade in values and custom overlaps for all your tracks, i can recommend you to set a big value (e.g 20 seconds) for Crossfade Duration in AutoDJ Settings. This way you will be sure the fade out duration will be equal with your custom overlaps duration. But that doesn't mean if you have a 5 seconds overlap, there will be no fade out, only because the default value in AutoDJ is 20 seconds. The fade out effect will still be appplied, but at least will be as long as the overlap duration is.

What i also recommend is using the **Bulk Media Import/Export** feature to make bulk changes for all that 0 fade in values. It's more easy to define that fade in fields in a .csv file, rather than editing every track in Media Editor. And then re-import the csv changes.

There's a second method of doing things:

- stick with the default value in AutoDJ Settings, the 2 seconds Crossfade Duration.
- set only overlaps for each and every track
- go to Export Media to CSV and export the csv file
- set faster the 0 fade in values in the csv file
- copy the column with overlap values and paste all that values to fade out column. this way all your fade out durations will be equal with overlap durations.
- Import Changes from CSV

Hope all this info will help you!

**Comment by unudelaradio** [comment]
Here's another method, which i think will fit better for you. This is the method i use and i like the most.

- go to AutoDJ Settings and Disable Crossfading
- set custom Cue Points in Visual Cue Editor, for all your songs & jingles
- paste the next custom code in the 3rd box of Liquidsoap Configuration

```
def clean_transition(old, new) =

  def safe_float_of_string(str) =
    try
      float_of_string(str)
    catch _ do
      0.
    end
  end

  cross_duration = safe_float_of_string(old.metadata["liq_cross_duration"])
  fade_out = safe_float_of_string(old.metadata["liq_fade_out"])
  fade_in = safe_float_of_string(new.metadata["liq_fade_in"])
  log_label = "clean_transition"
    
  let fade.out = fun (s) -> fade.out(type="exp", duration=fade_out, s)
  let fade.in = fun (s) -> fade.in(type="log", duration=fade_in, s)

  add = fun (a, b) -> add(normalize=false, [b, a])
  sequence = fun (a, b) -> (sequence([a, b]) : source)

  if cross_duration > 0. then
    log(label= log_label,"Crossfade. Duration: #{cross_duration}. Fade out: #{fade_out}. Fade in: #{fade_in}.")
    if fade_out > 0. and fade_in > 0. then 
      add(fade.out(old.source), fade.in(new.source))
    elsif fade_out > 0. and fade_in <= 0. then 
      add(fade.out(old.source), new.source)
    elsif fade_out <= 0. and fade_in > 0. then 
      add(old.source, fade.in(new.source))
    else
      add(old.source, new.source)
    end
  else
    log(label= log_label,"Sequence. Fade out: #{fade_out}. Fade in: #{fade_in}.")
    if fade_out > 0. and fade_in > 0. then 
      sequence(fade.out(old.source), fade.in(new.source))
    elsif fade_out > 0. and fade_in <= 0. then 
      sequence(fade.out(old.source), new.source)
    elsif fade_out <= 0. and fade_in > 0. then 
      sequence(old.source, fade.in(new.source))
    else
      sequence(old.source, new.source)
    end
  end

end

radio = cross(clean_transition,radio)
```

With this magic code created by RM-FM there will be no fade in and no fade out at all. There's no need to set 0 fade in values, and no fade out values, unless you want that for some specific tracks.

What i recommend you is to avoid long overlaps. Because Long Overlaps + Short Jingles = Issues. You can avoid long Overlaps by setting first the Cue Out Points. Setting a Cue Out closer to the Overlap point you want, will make the overlap shorter. 

Hope the code will work for you. I'm using it and it works perfectly with custom cue-points databases. And there are no fades, not even a second! With the code above, the only fades are the ones you set for some specific songs.

**Reply by wppronto** [reply]
Thank you so much!!!

**Comment by wppronto** [comment]
I'm having issues getting manual cue points to work reliably. This is without any custom LS code or anything.

I load some test songs in and set a cue out point (if needed), an overlap point and a fade-out point (the same value as you mentioned, for example 5 seconds). So the expected value would be for a song to play, with the next song starting at the overlap point, and the first song playing too for the specified fade-out period.

The problem is that some songs work and some don't, even with the same kinds of cue points set. Some fade properly, while others cut abruptly to the next song...no fade at all. Before I report it as a bug, I wanted to check whether it's user error or a bug.

**Reply by unudelaradio** [reply]
What are your AutoDJ settings in Azuracast? Crossfade Method and Crossfade Durarion (Seconds)?

**Reply by unudelaradio** [reply]
And what type of audio files do you use. I'm using mp3s only, with 128 to 320 kbps continuous bitrates. 

I mean... at some point there should be a difference between your settings and my settings. I'm curios where's the difference. If there's no difference between our working styles, there may be a bug. Or maybe some files and their types may cause this behaviour.

So let's recap your setup!

**Reply by wppronto** [reply]
I've tried smart and standard crossfades, leaving the default at 4 seconds.

Funny, I also wondered if file type made a difference. I'm listening and logging which ones work and which ones don't. I'll let you know if that is related.

**Comment by unudelaradio** [comment]
Maybe these screenshots will help you. This is how settings should look like, to avoid hard cuts.

1. The AutoDJ Settings. Setting 0.1 as general value helps you to avoid long _fade in_ effects at the start of the songs. Otherwise, if this general value is bigger, you should set 0 _fade in_ for each song in Media Editor.

![autodj](https://github.com/AzuraCast/AzuraCast/assets/74244061/1cf76fa3-8fe1-4014-9204-df984db3782e)

2.  Visual Cue Editor and Custom Cue Points:
 - Make sure the custom _fade out_ value is (at least) equal with the custom _overlap_ value.
 - Make sure you set custom _fade out_ values for all songs, otherwise, the _general_ value from AutoDJ Settings will be applied, both for _fade in_ and _fade out_ on all songs which don't have custom _fade in_ and _fade out_ values. 

![media](https://github.com/AzuraCast/AzuraCast/assets/74244061/86793462-67c9-49c8-96ad-102674bc6911)
